src/Polysemy/Final/Internal.hs

```diff
@@ -0,0 +1,211 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Polysemy.Final.Internal where
+
+import Data.Functor.Compose
+import Data.Maybe
+
+import Control.Concurrent.MVar
+import Control.Exception
+import Control.Monad
+import Control.Monad.Trans.Maybe
+
+import Polysemy
+import Polysemy.Internal
+import Polysemy.Internal.Union
+
+-----------------------------------------------------------------------------
+-- | An effect for embedding higher-order effects in the final target monad
+-- of the effect stack.
+--
+-- This is very useful for writing interpreters that interpret higher-order
+-- effects in terms of the final monad.
+--
+-- 'Final' is slightly weaker, but using it is otherwise more preferable
+-- than using a provided lowering function (a la 'Polysemy.Async.lowerAsync'):
+--
+--   * Interpreters using 'Final' may be run and composed like any other
+--     interpreter, unlike @lower-@ interpreters which need to be
+--     composed using '.@' or '.@@'.
+--
+--   * Multiple return-value changing interpreters using 'Final' may be
+--     composed together without issue. It's very difficult to
+--     compose 'Polysemy.Error.lowerError' with itself, unlike
+--     'Polysemy.Final.Error.errorToIOFinal'.
+--
+--   * Initialization work of @lower-@ interpreters may be duplicated
+--     when composed together with '.@'. @-'Final'@ interpreters avoid
+--     this issue altogether, as long as the initialization work is performed
+--     outside of a 'interpretH'/'interpretFinal'.
+--
+--   * Instead of having access to a natural transformation
+--     @forall x. 'Sem' r x -> m x@, @'Final' m@ provides a
+--     distribution through the effectful state:
+--     @forall x. f ('Sem' r x) -> m (f x)@,
+--     together with the initial state @f ()@, and an @'Inspector' f@.
+--     This is powerful enough for most purposes.
+--
+-- /Beware/: 'Final' actions are interpreted as actions of the final monad,
+-- and the effectful state visible to
+-- 'withWeaving'\/'Polysemy.Final.withStrategic'\/'Polysemy.Final.interpretFinal'
+-- is that of /all interpreters run in order to produce the final monad/.
+--
+-- This means that any interpreter built using 'Final' will /not/
+-- respect local/global state semantics based on the order of
```

Comment:
To concretize this for my understanding, this is essentially saying "`IORef`s are always global, so you shouldn't expect them to be local," right?

Reply:
oh nevermind, you explain immediately after!